### ISSUE TITLE
Can not get attribute from session sometimes

### ISSUE THREAD

**Issue by wangjohnny** [magro/memcached-session-manager#280]
software version: 
mom: 1.8.3, tomcat: 7.0.63, JDK: dk-8u51-linux-x64, OS: centos 6.6
I have 2 tomcats. A attribute is put into session in one tomcat(first access). But I can not find this attribute in this session in another tomcat(second access).
I found this problem occasionally. Not every time.

**Comment by magro** [comment]
Can you share your msm config please?

**Comment by wangjohnny** [comment]
context.xml:
    <Manager className="de.javakaffee.web.msm.MemcachedBackupSessionManager"
        memcachedNodes="memcached123.m.ocs.aliyuncs.com:11211"
        memcachedProtocol="binary"
        username="memcached123"
        password="123"
        sticky="true"
        sessionBackupAsync="true"
        sessionBackupTimeout="1000"
        requestUriIgnorePattern=".*\.(ico|png|gif|jpg|css|js)$"
        transcoderFactoryClass="de.javakaffee.web.msm.serializer.kryo.KryoTranscoderFactory"
        />

**Comment by RemyaPR** [comment]
Hi wangjohnny,
Is this issue resolved for you? We also face a similar problem.

**Comment by magro** [comment]
@wangjohnny Can you please try with sessionBackupAsync="false"?

**Comment by wangjohnny** [comment]
I try sessionBackupAsync="false" for several days. It looks like right! THKs

**Comment by magro** [comment]
@RemyaPR Does this solve the problem for you as well?

**Comment by RemyaPR** [comment]
hi @magro,

putting sessionBackupAsync="false" alone has not fixed our issue. But the frequency of the issue occurence had reduced with  sessionBackupAsync="false" setting. However, after putting the new jars you have provided (in the issue thread #282 ), the issue has not occurred again. I have updated this status in issue #282. Thank you for the support @magro. Your wonderful job has helped us to move the version to staging environment.

**Comment by wangjohnny** [comment]
@RemyaPR，
Since I use sessionBackupAsync="false", the issue dose not occur. But I'm still worrying about this issue, Because it occurs occasionally. I will adopt your solution. Thanks.
@magro ,
You always give me some help on time. Thanks for your support.

**Comment by felixup** [comment]
@magro ,

First of all, thank for what you done for this project . I also meet some problem when getting attribute from session.

## MSM Debug logs ：
```xml
详细: >>>>>> Request starting: POST //user/login (requestedSessionId CBFB2F02E5C1B5C1C24B5CEB0898501F-n1) ==================
一月 27, 2016 6:30:38 下午 de.javakaffee.web.msm.SessionIdFormat createSessionId
详细: Creating new session id with orig id 'ping' and memcached id 'n1'.
一月 27, 2016 6:30:38 下午 de.javakaffee.web.msm.NodeAvailabilityCache updateIsNodeAvailable
详细: CacheLoader returned node availability 'true' for node 'n1'.
一月 27, 2016 6:30:38 下午 de.javakaffee.web.msm.MemcachedSessionService loadFromMemcached
详细: Loading session from memcached: CBFB2F02E5C1B5C1C24B5CEB0898501F-n1
一月 27, 2016 6:30:38 下午 de.javakaffee.web.msm.LockingStrategyUriPattern onBeforeLoadFromMemcached
详细: Not lock request for request POST //user/login
一月 27, 2016 6:30:38 下午 de.javakaffee.web.msm.MemcachedSessionService loadFromMemcached
详细: Session CBFB2F02E5C1B5C1C24B5CEB0898501F-n1 not found in memcached.
一月 27, 2016 6:30:38 下午 de.javakaffee.web.msm.MemcachedSessionService createSession
详细: createSession invoked: CBFB2F02E5C1B5C1C24B5CEB0898501F-n1
一月 27, 2016 6:30:38 下午 de.javakaffee.web.msm.MemcachedSessionService createSession
详细: Created new session with id CBFB2F02E5C1B5C1C24B5CEB0898501F-n1
[2016-01-27 18:30:38.935] [http-bio-8081-exec-6] [INFO] [com.xbniao.mq.producer.XbnMessageProducer] - Begin to send message at [2016-01-27 18:30:38]
[2016-01-27 18:30:38.942] [http-bio-8081-exec-6] [INFO] [com.xbniao.mq.producer.XbnMessageProducer] - Success to send message at [2016-01-27 18:30:38]
[2016-01-27 18:30:38.943] [http-bio-8081-exec-6] [DEBUG] [com.xbniao.web.admin.controller.XbnAdminUserController] - ================================ LogIn ================================
[2016-01-27 18:30:38.961] [MessageSender-2] [INFO] [com.xbniao.mq.producer.MessageSender] - message : [key : 9304e02334da43f3a9beca5ea0b98fb4 , message : XbnMessageBody [id=9304e02334da43f3a9beca5ea0b98fb4, service=remoteXbnUserLoginLogService, method=save, obj=[XbnUserLoginLogDTO [id=null, userId=40289e5d500910e7015012668cbe0027, loginTime=Wed Jan 27 18:30:38 CST 2016, logoutTime=null, ip=192.168.30.43, ipLocation=null, createTime=null, updateTime=null, remark=ADMIN, isValid=null]], paramType=[class com.xbniao.usercenter.api.dto.XbnUserLoginLogDTO]]] has successfully sent to broker in 16 ms
一月 27, 2016 6:30:39 下午 de.javakaffee.web.msm.BackupSessionService backupSession
详细: Starting for session id CBFB2F02E5C1B5C1C24B5CEB0898501F-n1
一月 27, 2016 6:30:39 下午 de.javakaffee.web.msm.BackupSessionTask call
详细: Starting for session id CBFB2F02E5C1B5C1C24B5CEB0898501F-n1
一月 27, 2016 6:30:39 下午 de.javakaffee.web.msm.BackupSessionTask doBackupSession
详细: Trying to store session in memcached: CBFB2F02E5C1B5C1C24B5CEB0898501F-n1
一月 27, 2016 6:30:39 下午 de.javakaffee.web.msm.BackupSessionTask call
详细: Finished for session id CBFB2F02E5C1B5C1C24B5CEB0898501F-n1, returning status SUCCESS
一月 27, 2016 6:30:39 下午 de.javakaffee.web.msm.LockingStrategy onAfterBackupSession
详细: Stored session validity info for session CBFB2F02E5C1B5C1C24B5CEB0898501F-n1
一月 27, 2016 6:30:39 下午 de.javakaffee.web.msm.RequestTrackingHostValve logDebugRequestSessionCookie
详细: Have request session cookie: domain=null, maxAge=-1, path=null, value=CBFB2F02E5C1B5C1C24B5CEB0898501F-n1, version=0, secure=false
一月 27, 2016 6:30:39 下午 de.javakaffee.web.msm.RequestTrackingHostValve logDebugResponseCookie
详细: Request finished, with Set-Cookie header: JSESSIONID=CBFB2F02E5C1B5C1C24B5CEB0898501F-n1; Domain=.xbniao.com; Path=/; HttpOnly
一月 27, 2016 6:30:39 下午 de.javakaffee.web.msm.RequestTrackingHostValve invoke
详细: <<<<<< Request finished: POST //user/login ==================
[2016-01-27 18:30:39.964] [kafka-producer-network-thread | XBN-KFAKA-PRODUCER] [INFO] [com.xbniao.mq.producer.callback.XbnMessageCallBack] - message : XbnMessageBody [id=9304e02334da43f3a9beca5ea0b98fb4, service=remoteXbnUserLoginLogService, method=save, obj=[XbnUserLoginLogDTO [id=null, userId=40289e5d500910e7015012668cbe0027, loginTime=Wed Jan 27 18:30:38 CST 2016, logoutTime=null, ip=192.168.30.43, ipLocation=null, createTime=null, updateTime=null, remark=ADMIN, isValid=null]], paramType=[class com.xbniao.usercenter.api.dto.XbnUserLoginLogDTO]] has successfully sent to topic : APP_LOG_TOPIC , partition : 3, offset : 21589 in 1 s
一月 27, 2016 6:30:40 下午 de.javakaffee.web.msm.RequestTrackingHostValve invoke
详细: >>>>>> Request starting: GET //user/checkLogin?_=1453890653013 (requestedSessionId CBFB2F02E5C1B5C1C24B5CEB0898501F-n1) ==================
一月 27, 2016 6:30:40 下午 de.javakaffee.web.msm.SessionIdFormat createSessionId
详细: Creating new session id with orig id 'ping' and memcached id 'n1'.
一月 27, 2016 6:30:40 下午 de.javakaffee.web.msm.NodeAvailabilityCache updateIsNodeAvailable
详细: CacheLoader returned node availability 'true' for node 'n1'.
一月 27, 2016 6:30:40 下午 de.javakaffee.web.msm.MemcachedSessionService createSession
详细: createSession invoked: CBFB2F02E5C1B5C1C24B5CEB0898501F-n1
一月 27, 2016 6:30:40 下午 de.javakaffee.web.msm.MemcachedSessionService createSession
详细: Created new session with id CBFB2F02E5C1B5C1C24B5CEB0898501F-n1
一月 27, 2016 6:30:40 下午 de.javakaffee.web.msm.BackupSessionService backupSession
详细: Starting for session id CBFB2F02E5C1B5C1C24B5CEB0898501F-n1
一月 27, 2016 6:30:40 下午 de.javakaffee.web.msm.BackupSessionTask call
详细: Starting for session id CBFB2F02E5C1B5C1C24B5CEB0898501F-n1
一月 27, 2016 6:30:40 下午 de.javakaffee.web.msm.BackupSessionTask doBackupSession
详细: Trying to store session in memcached: CBFB2F02E5C1B5C1C24B5CEB0898501F-n1
一月 27, 2016 6:30:40 下午 de.javakaffee.web.msm.BackupSessionTask call
详细: Finished for session id CBFB2F02E5C1B5C1C24B5CEB0898501F-n1, returning status SUCCESS
一月 27, 2016 6:30:40 下午 de.javakaffee.web.msm.LockingStrategy onAfterBackupSession
详细: Stored session validity info for session CBFB2F02E5C1B5C1C24B5CEB0898501F-n1
一月 27, 2016 6:30:40 下午 de.javakaffee.web.msm.RequestTrackingHostValve logDebugRequestSessionCookie
详细: Have request session cookie: domain=null, maxAge=-1, path=null, value=CBFB2F02E5C1B5C1C24B5CEB0898501F-n1, version=0, secure=false
一月 27, 2016 6:30:40 下午 de.javakaffee.web.msm.RequestTrackingHostValve logDebugResponseCookie
详细: Request finished, with Set-Cookie header: JSESSIONID=CBFB2F02E5C1B5C1C24B5CEB0898501F-n1; Domain=.xbniao.com; Path=/; HttpOnly
一月 27, 2016 6:30:40 下午 de.javakaffee.web.msm.RequestTrackingHostValve invoke
详细: <<<<<< Request finished: GET //user/checkLogin?_=1453890653013 ==================
```


## Memcached logs ：
```xml
<36 get ping-n1
>36 END
<36 get CBFB2F02E5C1B5C1C24B5CEB0898501F-n1
>36 END
<36 set CBFB2F02E5C1B5C1C24B5CEB0898501F-n1 2050 7200 5265
>36 STORED
<36 set validity:CBFB2F02E5C1B5C1C24B5CEB0898501F-n1 2048 3600 20
>36 STORED
<36 get ping-n1
>36 END
<36 set CBFB2F02E5C1B5C1C24B5CEB0898501F-n1 2048 7200 89
>36 STORED
<36 set validity:CBFB2F02E5C1B5C1C24B5CEB0898501F-n1 2048 3600 20
>36 STORED
```

## Context.xml 
```xml
 <Manager className="de.javakaffee.web.msm.MemcachedBackupSessionManager"
     memcachedNodes="n1:192.168.30.210:11212"              
     sticky="false"  
     sessionBackupAsync="false"  
     lockingMode="uriPattern:/path1|/path2"  
     requestUriIgnorePattern=".*\.(ico|png|gif|jpg|css|js)$"  
     transcoderFactoryClass="de.javakaffee.web.msm.serializer.kryo.KryoTranscoderFactory"  
     />
```

##MSM VERSION 1.8.3

## Through the memcached log , i found msm set value to memcache when using request.getSession().getAttribute() method sometimes , maybe this is the reason. But i don't know why

**Comment by magro** [comment]
@wangjohnny @RemyaPR Great to hear it's working for you!

@felixup From the logs I can't see anything pointing to an issue with not getting attributes from the session. Not sure what exactly you mean with "i found msm set value to memcache when using request.getSession().getAttribute() method sometimes", please explain more detailed what you do, what you expect and what you see instead.
You could also try the jars mentioned in [#282](https://github.com/magro/memcached-session-manager/issues/282#issuecomment-170404984) and see if they fix the issue for you. I'll also release a version with these changes soon.

**Comment by magro** [comment]
I just released 1.9.1, should be fixed then.
